sklearn/metrics/pairwise.py

```diff
@@ -224,41 +234,156 @@ def euclidean_distances(X, Y=None, Y_norm_squared=None, squared=False,
+        return None, None
+    else:
+        if X_norm_squared is not None:
+            XX = np.atleast_1d(X_norm_squared).reshape(-1)
+            if XX.shape != (X.shape[0],):
+                raise ValueError(
+                    "Incompatible dimensions for X and X_norm_squared")
+        else:
+            XX = row_norms(X, squared=True)
+
+        if X is Y:  # shortcut in the common case euclidean_distances(X, X)
+            YY = XX
+        elif Y_norm_squared is not None:
+            YY = np.atleast_1d(Y_norm_squared).reshape(-1)
+            if YY.shape != (Y.shape[0],):
+                raise ValueError(
+                    "Incompatible dimensions for Y and Y_norm_squared")
+        else:
+            YY = row_norms(Y, squared=True)
+
+        XX = XX.astype(X.dtype, copy=False)
+        YY = YY.astype(Y.dtype, copy=False)
+
+        return XX, YY
+
+
+def _euclidean_distances_upcast_fast(X, XX, Y, YY):
+    """Euclidean distances between X and Y
+
+    Assumes X and Y have float32 dtype.
+    X and Y are upcasted to float64 by chunks, which size is chosen to limit
+    memory increase by approximately 10MiB.
+    """
+    n_samples_X = X.shape[0]
+    n_samples_Y = Y.shape[0]
+    n_features = X.shape[1]
+
+    distances = np.empty((n_samples_X, n_samples_Y), dtype=np.float32)
+
+    maxmem = 10 * 2**17  # this number of float64 take 10MiB memory.
+
+    x_density = X.getnnz() / np.prod(X.shape) if issparse(X) else 1
+    y_density = Y.getnnz() / np.prod(Y.shape) if issparse(Y) else 1
+
+    # The increase amount of memory is:
+    # - x_density * chunk_size * n_features (copy of chunk of X)
+    # - y_density * chunk_size * n_features (copy of chunk of Y)
+    # - chunk_size * chunk_size (chunk of distance matrix)
+    # Hence x² + (xd+yd)kx = M, where x=chunk_size, k=n_features, M=maxmem
+    #                                 xd=x_density and yd=y_density
+    tmp = (x_density + y_density) * n_features
+    chunk_size = (-tmp + np.sqrt(tmp**2 + 4 * maxmem)) / 2
+    chunk_size = max(int(chunk_size), 1)
+
+    n_samples_X_rem = n_samples_X % chunk_size
+    n_chunks_X = n_samples_X // chunk_size + (n_samples_X_rem != 0)
+    n_samples_Y_rem = n_samples_Y % chunk_size
+    n_chunks_Y = n_samples_Y // chunk_size + (n_samples_Y_rem != 0)
+
+    for i in range(n_chunks_X):
+        xs = i * chunk_size
+        xe = xs + (chunk_size if i < n_chunks_X - 1 else n_samples_X_rem)
+
+        X_chunk = X[xs:xe].astype(np.float64)
+        XX_chunk = row_norms(X_chunk, squared=True)
+
+        for j in range(n_chunks_Y):
+            ys = j * chunk_size
+            ye = ys + (chunk_size if j < n_chunks_Y - 1 else n_samples_Y_rem)
+
+            Y_chunk = Y[ys:ye].astype(np.float64)
+            YY_chunk = row_norms(Y_chunk, squared=True)
+
+            d = - 2 * safe_sparse_dot(X_chunk, Y_chunk.T, dense_output=True)
+            _add_norms(d, XX_chunk, YY_chunk)
+
+            distances[xs:xe, ys:ye] = d.astype(np.float32)
```

Comment:
Since we're chunking the matrices, why not handle the case where `X is Y` to compute only half of the matrix?
Is the code complexity not worth it?

Reply:
This is tricky because it's the final distance matrix which is symmetric, not the intermediate ones from chunking. It's possible of course but you can't use BLAS matrix matrix multiplication, leading to a huge drop of performances. Below is an illustration of why:

![image](https://user-images.githubusercontent.com/34657725/54205903-8a45d300-44d7-11e9-94ee-e9052d662500.png)

There's no way to only compute the elements of the matrix product below the line with BLAS.

Reply:
I wasn't suggesting the usage if `syrk`, but rather to compute only the chunks below the diagonal, and copy their transpose chunks above it.
Something like this.
```python
if X is Y:
    if i < j:
        distances[ys:ye, xs:xe] = distances[xs:se, ys:ye].T
    else:
        break
```
Then the only redundant computation would be for the chunks on the diagonal.

BTW, since your chunks are square, you can't have the diagonal of the final distance matrix cut the chunk anywhere else than its diagonal.

Meaning that for the chunks on the diagonal, the `syrk` trick could be used. And overall this would lead to exactly half the amount of computation when `X is Y`.

Reply:
You're right the fact that the chunk size is the same for X and Y makes it a lot easier. Your proposition seems simple enough and involves a very small number of unnecessary computations. I think it could be included in this PR.

Reply:
Actually you did that in your PR and I somehow didn't use it :(